Repl stack trace summary doesn't count suppressed frames correctly

## reproduction steps

```scala
scala> val e = new AssertionError("oops")
e: AssertionError = java.lang.AssertionError: oops

scala> e.addSuppressed(new java.io.IOException("bad io"))

scala> e.printStackTrace
java.lang.AssertionError: oops
        at $line3.$read$$iw$$iw$.<clinit>(<console>:1)
        at $line3.$eval$.$print$lzycompute(<synthetic>:6)
        at $line3.$eval$.$print(<synthetic>:5)
        at $line3.$eval.$print(<synthetic>)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at scala.tools.nsc.interpreter.IMain$ReadEvalPrint.call(IMain.scala:632)
        at scala.tools.nsc.interpreter.IMain$Request.loadAndRun(IMain.scala:946)
        at scala.tools.nsc.interpreter.IMain.$anonfun$interpret$1(IMain.scala:474)
        at scala.reflect.internal.util.ScalaClassLoader.asContext(ScalaClassLoader.scala:40)
        at scala.reflect.internal.util.ScalaClassLoader.asContext$(ScalaClassLoader.scala:37)
        at scala.reflect.internal.util.AbstractFileClassLoader.asContext(AbstractFileClassLoader.scala:40)
        at scala.tools.nsc.interpreter.IMain.loadAndRunReq$1(IMain.scala:473)
        at scala.tools.nsc.interpreter.IMain.$anonfun$interpret$3(IMain.scala:487)
        at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:487)
        at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:470)
        at scala.tools.nsc.interpreter.shell.ILoop.interpretStartingWith(ILoop.scala:931)
        at scala.tools.nsc.interpreter.shell.ILoop.command(ILoop.scala:788)
        at scala.tools.nsc.interpreter.shell.ILoop.processLine(ILoop.scala:462)
        at scala.tools.nsc.interpreter.shell.ILoop.loop(ILoop.scala:485)
        at scala.tools.nsc.interpreter.shell.ILoop.run(ILoop.scala:1020)
        at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:87)
        at scala.tools.nsc.MainGenericRunner.run$1(MainGenericRunner.scala:91)
        at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:103)
        at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:108)
        at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
        Suppressed: java.io.IOException: bad io
                at $line5.$read$$iw$$iw$.<clinit>(<console>:1)
                at $line5.$eval$.$print$lzycompute(<synthetic>:6)
                at $line5.$eval$.$print(<synthetic>:5)
                at $line5.$eval.$print(<synthetic>)
                ... 24 more

scala> throw e
java.lang.AssertionError: oops
  ... 28 elided
  Suppressed: java.io.IOException: bad io
    ... 24 more

scala>
```

## problem

```scala
scala> e.getSuppressed.head.printStackTrace
java.io.IOException: bad io
        at $line5.$read$$iw$$iw$.<clinit>(<console>:1)
        at $line5.$eval$.$print$lzycompute(<synthetic>:6)
        at $line5.$eval$.$print(<synthetic>:5)
        at $line5.$eval.$print(<synthetic>)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[snip]
```

## expectation

Printing the suppressed exception doesn't seem to count frames correctly, as the depth is 28. Here, the top four differ in the two exceptions.

--- a/src/compiler/scala/tools/nsc/util/StackTracing.scala
+++ b/src/compiler/scala/tools/nsc/util/StackTracing.scala
@@ -35,10 +35,10 @@ private[util] trait StackTracing extends Any {
     val Suppressed = new TraceRelation("Suppressed: ")
 
     def header(e: Throwable): String  = {
-      def because(e: Throwable): String = e.getCause match { case null => null ; case c => header(c) }
-      def msg(e: Throwable): String     = e.getMessage match { case null => because(e) ; case s => s }
-      def txt(e: Throwable): String     = msg(e) match { case null => "" ; case s => s": $s" }
-      s"${e.getClass.getName}${txt(e)}"
+      def because = e.getCause   match { case null => null    ; case c => header(c) }
+      def msg     = e.getMessage match { case null => because ; case s => s         }
+      def txt     = msg          match { case null => ""      ; case s => s": $s"   }
+      s"${e.getClass.getName}$txt"
     }
 
     val seen = mutable.Set.empty[Throwable]
@@ -54,16 +54,22 @@ private[util] trait StackTracing extends Any {
     def print(e: Throwable, r: TraceRelation, share: Array[StackTraceElement], indents: Int): Unit = if (unseen(e)) {
       val trace  = e.getStackTrace
       val frames = if (share.isEmpty) trace else {
-        val spare  = share.reverseIterator
+        val spare   = share.reverseIterator
         val trimmed = trace.reverse dropWhile (spare.hasNext && spare.next == _)
         trimmed.reverse
       }
       val prefix = frames.takeWhile(p)
       val margin = "  " * indents
-      lines += s"${margin}${r}${header(e)}"
-      prefix.foreach(f => lines += s"${margin}  at $f")
-      if (frames.size < trace.size) lines += s"${margin}  ... ${trace.size - frames.size} more"
-      if (r == Self && prefix.size < frames.size) lines += s"${margin}  ... ${frames.size - prefix.size} elided"
+      lines += s"$margin$r${header(e)}"
+      prefix.foreach(frame => lines += s"$margin  at $frame")
+
+      val traceFramesLenDiff  = trace.length - frames.length
+      val framesPrefixLenDiff = frames.length - prefix.length
+      if (traceFramesLenDiff > 0) {
+        if (framesPrefixLenDiff > 0) lines += s"$margin  ... $framesPrefixLenDiff elided and $traceFramesLenDiff more"
+        else lines += s"$margin  ... $traceFramesLenDiff more"
+      } else if (framesPrefixLenDiff > 0) lines += s"$margin  ... $framesPrefixLenDiff elided"
+
       print(e.getCause, CausedBy, trace, indents)
       e.getSuppressed.foreach(print(_, Suppressed, frames, indents + 1))
     }

--- a/src/partest/scala/tools/partest/ReplTest.scala
+++ b/src/partest/scala/tools/partest/ReplTest.scala
@@ -81,10 +81,12 @@ trait StackCleaner extends ReplTest {
   override def normalize(s: String) = stripFrameCount(super.normalize(s))
 }
 object StackCleaner {
-  private val elided = """(\s+\.{3} )\d+( elided)""".r
-  private val frame  = """(\s+at [^(]+\(<console>:)\d+(\))""".r
+  private val elidedAndMore = """(\s+\.{3} )\d+( elided and )\d+( more)""".r
+  private val elidedOrMore        = """(\s+\.{3} )\d+( (?:elided|more))""".r
+  private val frame         = """(\s+at [^(]+\(<console>:)\d+(\))""".r
   private def stripFrameCount(line: String) = line match {
-    case elided(ellipsis, suffix) => s"$ellipsis???$suffix"
+    case elidedAndMore(ellipsis, infix, suffix) => s"$ellipsis???$infix???$suffix" // must be before `elided`
+    case elidedOrMore(ellipsis, suffix) => s"$ellipsis???$suffix"
     case frame(prefix, suffix)    => s"${prefix}XX${suffix}"
     case s                        => s
   }

--- a/test/files/run/repl-trace-elided-more.check
+++ b/test/files/run/repl-trace-elided-more.check
@@ -1,0 +1,25 @@
+
+scala> val e = new AssertionError("oops")
+val e: AssertionError = java.lang.AssertionError: oops
+
+scala> e.addSuppressed(new java.io.IOException("bad io"))
+
+scala> throw e
+java.lang.AssertionError: oops
+  ... ??? elided
+  Suppressed: java.io.IOException: bad io
+    ... ??? elided and ??? more
+
+scala> val f = new Exception("f"); val g = new Exception("g")
+val f: Exception = java.lang.Exception: f
+val g: Exception = java.lang.Exception: g
+
+scala> f.addSuppressed(g)
+
+scala> throw f
+java.lang.Exception: f
+  ... ??? elided
+  Suppressed: java.lang.Exception: g
+    ... ??? more
+
+scala> :quit

--- a/test/files/run/repl-trace-elided-more.scala
+++ b/test/files/run/repl-trace-elided-more.scala
@@ -1,0 +1,5 @@
+import scala.tools.partest.{SessionTest, StackCleaner}
+
+// scala/bug#11945 print stack trace elided count correctly
+//
+object Test extends SessionTest with StackCleaner

